### ISSUE TITLE
REQUEST: Disable auto-retry upon any failure/error

### ISSUE THREAD

**Issue by Jon8RFC** [icing/mod_md#222]
An option to disable any and all automatic retries, for anything other than a mod_md perfect success.

By the time I can see that it failed, it's already retried at least once.  The limit for letsencrypt staging is 60 failures per hour (only 5 for production) with other restrictions in tandem.  As best I can tell, letsencrypt doesn't specify "you're rate limited", and instead fails without any notification and mod_md continues retrying, regardless.  This makes troubleshooting quite difficult when each apache restart ends up retrying multiple times when I want to test just once to narrow down what the issue might be.

I'm still trying to figure out why my certs won't renew (for the past month and a half), and time is running out, haha.

**Comment by Jon8RFC** [comment]
Side note:
Thankfully, the archived certs (thank you!) let me look far back enough to see that at one point it setup challenges as http-01, and in all of the failures I've seen tls-alpn-01.

There was always success with staging, but 100% failure with production letsencrypt renewals and no helpful failure information other than "internal failure".  Even on recent successes with cert renewals, the type of challenge wasn't logged for some reason, so I never knew those successes used anything different.  I had to go back many, many months to find a log that showed http-01 which helped me realize that it was something new to try.  Setting MDCAChallenges http-01 in httpd.conf resolved the issue which I have no idea how it began or why letsencrypt staging worked flawlessly.

It always said "internal error" and there was nothing useful in the apache log or job.json:
"status": 20014,
    "status-description": "Internal error (specific information not available)",
    "detail": "Unable to retrieve rsa certificate chain.",

It was the challenge setup that was failing, so hopefully this will be useful for internet searches or gives you more info to make a new error message.

**Comment by icing** [comment]
@Jon8RFC that is good to know. I have to reproduce this failure and check the messages/logs the module gives here. It should be able to

  a. be more helpful in informing about the cause
  b. stop retrying if the problem can only be fixes by a change in server configuration

**Comment by Jon8RFC** [comment]
I'm not comfortable publicly posting my configs, but if there's an email I can send them to, I'm happy to send my entire apache configs if that's helpful.

**Comment by icing** [comment]
This issue remains open and is also part of the (now fixed) #227. The basic question is when mod_md shall retry and when/if it shall abstain. And it's no easy answer.

I guess what I can see is that the module carries a list of CA error codes that indicates a failure with the input and not something temporary. For theses errors, it does not make sense to retry again in a couple of minutes.

However, even  CAs can have bugs and a seemingly permanent error code might be fixed on the next day. So, I think that a list of error codes that immediately switches to the largest retry interval (daily) would make most sense. And retain the perseverance to live through error conditions.

**Comment by Jon8RFC** [comment]
That sounds great.  From what I remember, hiccups at Let's Encrypt were rarely longer than a day, so a 1 day interval is excellent and avoids rate-limiting, especially since the default mod_md renewal has a 1-month buffer to work with.

**Comment by icing** [comment]
@Jon8RFC coming back to your original issue. If I understand correctly, you had set "tls-alpn-01" as challenge and that let to the "internal error" that gave you no hint as to what was wrong?

**Comment by Jon8RFC** [comment]
@icing I didn't have the challenge manually set (it was doing whatever was default), but when looking back at much older logs it did show which challenge type it used and I could see that it was normally succeeding with http-01 in the past, and old failures with tls-alpn-01 (it alternated on its own and I never set either one), but I don't recall the situation of those failures, and it may have been that the older apache versions cooperated better with the newer mod_md.  When I manually set it to http-01, then things worked.


One issue was that I didn't know that the internal error was related to the challenge in those failures.  The logs didn't indicate that it failed at the challenge like the much older logs did, for some reason, so I'm wondering if it was related to older apache (2.4.43 or older) being more compliant than 2.4.43a - 2.4.46.  Or, possibly, it was a mismatch of openssl when I manually updated and didn't know what I was doing?  I'll keep this in mind with the next openssl update and try manually updating.

Another issue was that the logs didn't show the failure OR success of the challenges in the logs (during the time of things not working), just that it worked.  In those logs where it succeeded, it didn't show which challenge type it used.  Only when looking back at much older logs did it log the challenge type successes and failures.

I do see the challenges showing up now, but the "Setting up challenge 'tls-alpn-01' for domain jon8rfc.homeip.net" is what was missing in both the successes and the failures.  When it failed, it jumped to "Monitoring challenge status for jon8rfc.homeip.net", so when it failed, I didn't know it failed at tls-alpn-01 because it wasn't showing that in the successes or failures, for some reason.

**Comment by icing** [comment]
So, the differences came when switching from the mod_md in 2.4.43 (or older) to the 2.4.46?

**Comment by Jon8RFC** [comment]
I'm wondering if maybe it was apache-related, but I carried over the same 2.3.3-beta mod_md so the variable would be apache rather than mod_md.
Or, possibly my manual updates of openssl causing some strange mismatch but manually updating didn't cause issues with mod_md or my site in other instances.  Unfortunately, I cleaned out old stuff now that things were resolved so I can't check for specific dates.
